Specifications/Language/3_Expressions/Closures.md

```diff
@@ -0,0 +1,118 @@
+# Closures
+
+Closures are callables that capture variables from the enclosing environment.
+Both function and operation closures can be created.
+An operation closure can be created inside a function, but it can only be applied in an operation.
+
+Q# has two mechanisms for creating closures: lambda expressions and partial application.
+
+## Lambda Expressions
+
+A lambda expression creates an anonymous function or operation.
+The basic syntax is a symbol tuple to bind the parameters, an arrow (`->` for a function and `=>` for an operation), and an expression to evaluate when applied.
+
+```qsharp
+// Function that captures 'x':
+y -> x + y
+
+// Operation that captures 'qubit':
+deg => Rx(deg * PI() / 180.0, qubit)
+
+// Function that captures nothing:
+(x, y) -> x + y
+```
+
+### Parameters
+
+Parameters are bound using a symbol tuple that is identical to the left-hand side of a [variable declaration statement](../2_Statements/VariableDeclarationsAndReassignments.md).
+The type of the parameter tuple is implicit.
+Type annotations are not supported; if type inference fails, you may need to create a top-level callable declaration and use partial application instead.
+
+### Mutable Capture Variables
+
+Mutable variables cannot be captured.
+If you only need to capture the value of a mutable variable at the instant the lambda expression is created, you can create an immutable copy:
+
+```qsharp
+// ERROR: 'variable' cannot be captured.
+mutable variable = 1;
+let f = () -> variable;
+
+// OK.
+let value = variable;
+let g = () -> value;
+```
+
+### Characteristics
+
+The characteristics of an anonymous operation are inferred based on the body of the lambda expression.
+For example:
+
+```qsharp
+// Has type Unit => Unit is Adj + Ctl because X is known to be Adj + Ctl.
+() => X(q)
+
+
+// Has type Unit => Result because M is neither Adj nor Ctl.
+() => M(q)
+```
+
+Because of limitations in characteristics inference, this is based only on type information known by the point where the lambda expression occurs.
+For example:
+
+```qsharp
+let foo = op => op(q);
+foo(X);
+```
+
+Here, `foo` is inferred to have type:
+
+```qsharp
+(Qubit => Unit is Adj + Ctl) => Unit
+```
+
+But that type is *not* the principal type for `foo`, which is:
```

Comment:
> I think a short description of "principal type" would be helpful.

I agree. This technical term is also used in #140, and I also think it might be useful to have some explanation.

About the type of `foo`: since `op` gets the type of `X`, i.e., its argument in `foo(X)`, the _most specific type_ that can be inferred is `(Qubit => Unit is Adj + Ctl) => Unit is Adj + Ctl` because it is used in both the input and the output of that lambda.

But it raises the question of the meaning of "principal type" in this context, as it usually means _the most general type_. Alan's suggestion should be the answer when we do not know what `op` may get instantiated to.

Reply:
Ah -- right -- I missed the implied inference :-) based on the type of `X`, I was just thinking of `foo` based on it's definition.

Reply:
I changed the wording to clarify that `foo`'s type also depends on `X`'s type. I also replaced "principal type" with "most specific type" to avoid having to define it.

> But it raises the question of the meaning of "principal type" in this context, as it usually means the most general type.

I don't think principal type necessarily means most general type - just a type that can be instantiated into every other possible type for that term. For subtyping systems, I think the most specific type would count, since it can be cast into every other (more general) type. Maybe there is another nuance for subtypes that I don't know about though.